### PR TITLE
[Feature Request] Inserting or Deleting Table Rows

### PR DESCRIPTION
**Pull request by danthorpe**
Add the ability to for the TaylorSource datasource to handle UITableView insertions and deletions:
```objc
// Insert or delete table rows
- tableView:commitEditingStyle:forRowAtIndexPath:

// Ask if the given row is editable
- tableView:canEditRowAtIndexPath:
```

### CODE REVIEW

**Comment by danthorpe** [comment]
Okay, so... I modified the approach above to use optional closures on `DatasourceProviderType` instead. 

```swift
public protocol DatasourceProviderType {
    typealias Datasource: DatasourceType

    var datasource: Datasource { get }

    var canEditItemAtIndexPath: CanEditItemAtIndexPath? { get }
    var commitEditActionForItemAtIndexPath: CommitEditActionForItemAtIndexPath? { get }
    var canMoveItemAtIndexPath: CanMoveItemAtIndexPath? { get }
    var commitMoveItemAtIndexPathToIndexPath: CommitMoveItemAtIndexPathToIndexPath? { get }

    var editActionForItemAtIndexPath: EditActionForItemAtIndexPath? { get }
}
```

I think this is the most elegant solution, and is forwards compatible with Swift 2.0 which will be able to provide a default `= .None` implementation.

The only requirement is that the group of 4 closures must all be implemented to get a table view datasource which implements the corresponding optional methods. The 5th closure is a helper so that the logic for editing can be in the same place - your table view delegate (likely the view controller) can then access this.

Take a look at `ViewController.swift` in the Events example for this being used. Note that it works nicely with the `SegmentedDatasourceProvider`.

@king7532 if you want to try this in your project, you can build against it using this in your Podfile

```ruby
pod 'TaylorSource', :git => 'https://github.com/danthorpe/TaylorSource.git', :branch => 'feature/TAY-18_subclassable_table_view_datasource'
```

I'll not merge it until I've written unit tests around these changes - but, hopefully should get that done in the next couple of days.

File: examples/Gallery/Gallery/ViewController.swift [hunk 1]
```diff
@@ -190,6 +190,12 @@ class GalleryDatasource: DatasourceProviderType {
     var datasource: Datasource
     var date: NSDate
 
+    var canEditItemAtIndexPath: CanEditItemAtIndexPath? = .None
+    var commitEditActionForItemAtIndexPath: CommitEditActionForItemAtIndexPath? = .None
+    var editActionForItemAtIndexPath: EditActionForItemAtIndexPath? = .None
+    var canMoveItemAtIndexPath: CanMoveItemAtIndexPath? = .None
+    var commitMoveItemAtIndexPathToIndexPath: CommitMoveItemAtIndexPathToIndexPath? = .None
+
```
Comment (danthorpe) [inline examples/Gallery/Gallery/ViewController.swift]: This won't be necessary in Swift 2.0 version.

**Comment by king7532** [comment]
@danthorpe Strong work!  That's what I was hoping for and thank you for implementing it in one of the example projects!  It looks like I can easily add it to my project.

**Comment by danthorpe** [comment]
@king7532 no problem - would be great to know how you get on with it!

**Comment by danthorpe** [comment]
@king7532 as this has now been merged into `development`, until version `0.14` is available in CocoaPods, you can build the development branch directly in your project with the following line in your Podfile:

```ruby
pod 'TaylorSource/YapDatabase', :git => 'https://github.com/danthorpe/TaylorSource.git', :branch => 'development'
```

**Comment by king7532** [comment]
@danthorpe Thank you!  (FYI, I know CocoaPods and git very well ;)

**Comment by danthorpe** [comment]
@king7532 okay, great.

**Pull request merged**